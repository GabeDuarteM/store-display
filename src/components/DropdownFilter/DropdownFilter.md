```js
initialState = {
  menuIsOpen: false,
  selectedItem: { id: 'price-asc', text: 'Price: low to high' },
}
;<DropdownFilter
  items={[
    { id: 'price-asc', text: 'Price: low to high' },
    { id: 'price-desc', text: 'Price: high to low' },
    { id: 'rating-asc', text: 'Rating: low to high' },
    { id: 'rating-desc', text: 'Rating: high to low' },
  ]}
  selectedItem={state.selectedItem}
  onChange={selectedItem => setState({ selectedItem })}
  menuIsOpen={state.menuIsOpen}
  handleChangeMenuIsOpen={menuIsOpen => {
    setState({ menuIsOpen })
  }}
  label="Sort by"
/>
```
