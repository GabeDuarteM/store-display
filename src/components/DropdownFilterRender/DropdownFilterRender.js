import React from 'react'

import PropTypes from 'prop-types'
import injectSheet from 'react-jss'

import DropdownFilterInput from '../DropdownFilterInput'
import DropdownFilterItemsList from '../DropdownFilterItemsList'

export const styles = {
  renderRoot: {
    display: 'flex',
    alignItems: 'center',
    cursor: 'default',
    userSelect: 'none',
  },
  label: {
    fontSize: 12,
    color: '#585858',
    fontWeight: 500,
    marginRight: 8,
  },
}

const DropdownFilterRender = ({
  classes,
  label,
  getLabelProps,
  getButtonProps,
  getItemProps,
  isOpen,
  handleChangeMenuIsOpen,
  selectedItem,
  width,
  items,
}) => (
  <div className={classes.renderRoot}>
    {label && (
      <label className={classes.label} {...getLabelProps()}>
        {label}
      </label>
    )}
    <div>
      <DropdownFilterInput
        getButtonProps={getButtonProps}
        isOpen={isOpen}
        handleChangeMenuIsOpen={handleChangeMenuIsOpen}
        selectedItem={selectedItem}
        width={width}
      />
      {isOpen ? (
        <DropdownFilterItemsList
          items={items}
          getItemProps={getItemProps}
          selectedItem={selectedItem}
          width={width}
        />
      ) : null}
    </div>
  </div>
)

export const propTypes = {
  /** @ignore */
  classes: PropTypes.object.isRequired,
  /** If provided, it displays a label to the dropdown */
  label: PropTypes.string,
  /** Function to get the props from Downshift */
  getLabelProps: PropTypes.func.isRequired,
  /** Function to get the props from Downshift */
  getButtonProps: PropTypes.func.isRequired,
  /** Function to get the props from Downshift */
  getItemProps: PropTypes.func.isRequired,
  /** Defines if the dropdown menu should be open */
  isOpen: PropTypes.bool.isRequired,
  /**
   * Gets called when the dropdown gets or loses focus
   * @param {bool} menuIsOpen The value that defines if the menu should or should not be open
   */
  handleChangeMenuIsOpen: PropTypes.func.isRequired,
  /** The current selected item to be displayed on the input */
  selectedItem: PropTypes.shape({
    id: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
  }).isRequired,
  /** The width of the component */
  width: PropTypes.number.isRequired,
  /** Array of items to display */
  items: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      text: PropTypes.string.isRequired,
    }),
  ).isRequired,
}

export const defaultProps = {
  label: null,
}

DropdownFilterRender.propTypes = propTypes
DropdownFilterRender.defaultProps = defaultProps

export default injectSheet(styles)(DropdownFilterRender)
