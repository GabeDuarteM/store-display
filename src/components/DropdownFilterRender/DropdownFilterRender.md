```js
<div style={{ height: 155 }}>
  <DropdownFilterRender
    label="Label"
    getLabelProps={customProps => customProps}
    getButtonProps={customProps => customProps}
    getItemProps={customProps => customProps}
    isOpen
    handleChangeMenuIsOpen={() => {}}
    selectedItem={{ id: 'price-asc', text: 'item 1' }}
    width={200}
    items={[
      { id: 'price-asc', text: 'item 1' },
      { id: 'price-desc', text: 'item 2' },
      { id: 'item-3', text: 'item 3' },
    ]}
  />
</div>
```
