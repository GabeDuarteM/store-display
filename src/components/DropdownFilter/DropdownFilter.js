import React from 'react'

import PropTypes from 'prop-types'
import Downshift from 'downshift'

import DropdownFilterRender from '../DropdownFilterRender'

const DropdownFilter = ({
  items,
  label,
  onChange,
  menuIsOpen,
  handleChangeMenuIsOpen,
  selectedItem,
  width,
}) => (
  <Downshift
    onChange={onChange}
    onSelect={() => handleChangeMenuIsOpen(false)}
    selectedItem={selectedItem}
    isOpen={menuIsOpen}
    onOuterClick={() => {
      handleChangeMenuIsOpen(false)
    }}
    itemToString={item => item.text}
    render={({
      getButtonProps,
      getLabelProps,
      getItemProps,
      isOpen,
      selectedItem: selectedItemRender,
    }) => (
      <div>
        <DropdownFilterRender
          label={label}
          getLabelProps={getLabelProps}
          getButtonProps={getButtonProps}
          getItemProps={getItemProps}
          isOpen={isOpen}
          handleChangeMenuIsOpen={handleChangeMenuIsOpen}
          selectedItem={selectedItemRender}
          width={width}
          items={items}
        />
      </div>
    )}
  />
)

export const propTypes = {
  /** Array of items to display */
  items: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      text: PropTypes.string.isRequired,
    }),
  ).isRequired,
  /** Gets called when the user changes the selected item */
  onChange: PropTypes.func.isRequired,
  /**
   * Gets called when the dropdown gets or loses focus
   * @param {bool} menuIsOpen The value that defines if the menu should or should not be open
   */
  handleChangeMenuIsOpen: PropTypes.func.isRequired,
  /** The current selected item */
  selectedItem: PropTypes.shape({
    id: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
  }).isRequired,
  /** The width of the component */
  width: PropTypes.number,
  /** If provided, it displays a label to the dropdown */
  label: PropTypes.string,
  /** Defines if the dropdown menu should be open */
  menuIsOpen: PropTypes.bool,
}

export const defaultProps = {
  width: 200,
  label: null,
  menuIsOpen: false,
}

DropdownFilter.propTypes = propTypes
DropdownFilter.defaultProps = defaultProps

export default DropdownFilter
