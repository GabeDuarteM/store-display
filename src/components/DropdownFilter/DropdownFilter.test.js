import React from 'react'

import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import DropdownFilter, { propTypes, defaultProps } from './DropdownFilter'
import DropdownFilterRender from '../DropdownFilterRender'

const createWrapper = (func, optionalProps) =>
  func(
    <DropdownFilter
      items={[
        { id: 'item-1', text: 'item 1' },
        { id: 'item-2', text: 'item 2' },
        { id: 'item-3', text: 'item 3' },
      ]}
      selectedItem={{ id: 'item-2', text: 'item 2' }}
      itemToString={item => item.text}
      onChange={jest.fn}
      handleChangeMenuIsOpen={jest.fn}
      {...optionalProps}
    />,
  )

const createWrapperShallow = optionalProps =>
  createWrapper(shallow, optionalProps)

const createWrapperMount = optionalProps => createWrapper(mount, optionalProps)

describe('COMPONENT: <DropdownFilter />', () => {
  it('should render without crashing with default props', () => {
    createWrapperMount()
  })

  it('should render without crashing with custom props', () => {
    createWrapperMount({ label: 'Something' })
  })

  it("should match component's snapshot", () => {
    const wrapper = createWrapperShallow()
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it("should match component's propTypes snapshot", () => {
    expect(propTypes).toMatchSnapshot()
  })

  it("should match component's defaultProps snapshot", () => {
    expect(defaultProps).toMatchSnapshot()
  })

  it('should call handleChangeMenuIsOpen passing false as an argument when onSelect is called', () => {
    const handleChangeMenuIsOpen = jest.fn()
    const wrapper = createWrapperShallow({ handleChangeMenuIsOpen })

    const onSelect = wrapper.prop('onSelect')
    onSelect()

    expect(handleChangeMenuIsOpen).toHaveBeenCalledTimes(1)
    expect(handleChangeMenuIsOpen).toHaveBeenLastCalledWith(false)
  })

  it('should call handleChangeMenuIsOpen passing false as an argument when onOuterClick is called', () => {
    const handleChangeMenuIsOpen = jest.fn()
    const wrapper = createWrapperShallow({ handleChangeMenuIsOpen })

    const onOuterClick = wrapper.prop('onOuterClick')
    onOuterClick()

    expect(handleChangeMenuIsOpen).toHaveBeenCalledTimes(1)
    expect(handleChangeMenuIsOpen).toHaveBeenLastCalledWith(false)
  })

  it('should have the render method returning a DropdownFilterRender component', () => {
    const wrapper = createWrapperShallow()
    const render = wrapper.prop('render')
    const wrapperRender = shallow(
      render({
        getButtonProps: customProps => customProps,
        getLabelProps: customProps => customProps,
        getItemProps: customProps => customProps,
        isOpen: true,
        selectedItem: 'item 1',
      }),
    )
    const dropdownFilterRenderExists = wrapperRender
      .find(DropdownFilterRender)
      .exists()

    expect(dropdownFilterRenderExists).toBeTruthy()
  })

  it("should return the item's text when itemToString is called passing an item", () => {
    const item = { id: 'item-2', text: 'Item 2' }
    const wrapper = createWrapperShallow()
    const itemToString = wrapper.prop('itemToString')

    expect(itemToString(item)).toBe(item.text)
  })
})
