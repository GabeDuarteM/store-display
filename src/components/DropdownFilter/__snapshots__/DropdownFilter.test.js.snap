// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT: <DropdownFilter /> should match component's defaultProps snapshot 1`] = `
Object {
  "label": null,
  "menuIsOpen": false,
  "width": 200,
}
`;

exports[`COMPONENT: <DropdownFilter /> should match component's propTypes snapshot 1`] = `
Object {
  "handleChangeMenuIsOpen": [Function],
  "items": [Function],
  "label": [Function],
  "menuIsOpen": [Function],
  "onChange": [Function],
  "selectedItem": [Function],
  "width": [Function],
}
`;

exports[`COMPONENT: <DropdownFilter /> should match component's snapshot 1`] = `
<Downshift
  breakingChanges={Object {}}
  defaultHighlightedIndex={null}
  defaultInputValue=""
  defaultIsOpen={false}
  defaultSelectedItem={null}
  environment={[Window]}
  getA11yStatusMessage={[Function]}
  isOpen={false}
  itemToString={[Function]}
  onChange={[Function]}
  onInputValueChange={[Function]}
  onOuterClick={[Function]}
  onSelect={[Function]}
  onStateChange={[Function]}
  onUserAction={[Function]}
  render={[Function]}
  selectedItem={
    Object {
      "id": "item-2",
      "text": "item 2",
    }
  }
  selectedItemChanged={[Function]}
  stateReducer={[Function]}
/>
`;
